lib/async-emitter.js

```diff
@@ -0,0 +1,106 @@
+'use strict';
+
+class AsyncEmitter {
+  constructor() {
+    this.events = new Map();
+  }
+
+  // Get or create event
+  //   name <string> event name
+  // Returns: { on: <Set>, once: <Set> } }
+  event(name) {
+    const { events } = this;
+    const event = events.get(name);
+    if (event) return event;
+    const res = { on: new Set(), once: new Set() };
+    events.set(name, res);
+    return res;
+  }
+
+  // Add listener
+  //   name <string> event name
+  //   fn <Function> listener
+  on(name, fn) {
+    this.event(name).on.add(fn);
+  }
+
+  // Add listener
+  //   name <string> event name
+  //   fn <Function> listener
+  // Returns: <Promise> | <null>
+  once(name, fn) {
+    if (fn === undefined) {
+      return new Promise(resolve => {
+        this.once(name, resolve);
+      });
+    }
+    this.event(name).once.add(fn);
+    return Promise.resolve();
+  }
+
+  // Emit event
+  //   name <string> event name
+  //   args <any[]>
+  // Returns: <Promise>
+  emit(name, ...args) {
+    const { events } = this;
+    const event = events.get(name);
+    if (!event) return Promise.resolve();
+    const { on, once } = event;
+    const promises = [...on, ...once].map(fn => fn(...args));
+    once.clear();
```

Comment:
I'd prefer to move this one line up, because some of the listeners may create additional `once` listeners.